### PR TITLE
gcsio: reduce number of get requests in function calls

### PR DESCRIPTION
**Pull request by shunping**
We replace `client.get_bucket()` with `client.bucket()`, and `bucket.get_blob()` with `bucket.blob()` in a few places where a lightweight instance of bucket/blob is sufficient for the function calls.

The fix reduces unnecessary get requests (from `get_bucket` and `get_blob`) in the following gcsio apis:
* open
* delete
* delete_batch
* copy
* copy_batch
* list_files

addresses #28398

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Review by lgeiger — COMMENTED** [review]
Should [`.get_bucket` in `self._gcs_object`](https://github.com/apache/beam/blob/d7f7b43248b370188df6775b604b70fd63512713/sdks/python/apache_beam/io/gcp/gcsio.py#L418) also be replaced with `.bucket`?

It seems like [`bucket.get_blob`](https://github.com/googleapis/python-storage/blob/a0416a24c6194850177082bec023346b8acd17a6/google/cloud/storage/bucket.py#L1176-L1281) already seems to make a HTTP request to check whether the blob exists, or am I missing something?

**Comment by shunping** [comment]
> Should [`.get_bucket` in `self._gcs_object`](https://github.com/apache/beam/blob/d7f7b43248b370188df6775b604b70fd63512713/sdks/python/apache_beam/io/gcp/gcsio.py#L418) also be replaced with `.bucket`?
> 
> It seems like [`bucket.get_blob`](https://github.com/googleapis/python-storage/blob/a0416a24c6194850177082bec023346b8acd17a6/google/cloud/storage/bucket.py#L1176-L1281) already makes a HTTP request to check whether the blob exists, or am I missing something?

I checked the code, it should be fine to replace the `get_bucket()` with `bucket()` within `_gcs_object()`.

**Comment by shunping** [comment]
assign to next reviewer

File: sdks/python/apache_beam/io/gcp/gcsio.py [hunk 1]
```diff
@@ -228,16 +225,15 @@ def delete_batch(self, paths):
       with current_batch:
         for path in current_paths:
           bucket_name, blob_name = parse_gcs_path(path)
-          bucket = self.client.get_bucket(bucket_name)
+          bucket = self.client.bucket(bucket_name)
           bucket.delete_blob(blob_name)
 
       for i, path in enumerate(current_paths):
         error_code = None
-        for j in range(2):
```
Comment (Abacn) [inline sdks/python/apache_beam/io/gcp/gcsio.py:236]: Good point. That was why it needed to be range(2) or range(4) in delete_batch and copy_batch

**Comment by johnjcasey** [comment]
LGTM

**Comment by Abacn** [comment]
Checked that tests passed and Comments are resolved, merging for now

**Comment by Abacn** [comment]
seems this change need to make test change accordingly. PythonPostCommit failing `apache_beam.io.gcp.gcsfilesystem_integration_test.GcsFileSystemIntegrationTest.test_rename_error ` failing as it was on the premise of the request of delete one file returning two response code CC: @shunping 

https://github.com/apache/beam/runs/21333859365

**Pull request merged**